libraries/lib-sentry-reporting/SentryReport.cpp

```diff
@@ -0,0 +1,430 @@
+{
+   rapidjson::Value messageObject(rapidjson::kObjectType);
+
+   messageObject.AddMember(
+      "message",
+      rapidjson::Value(
+         message.Value.c_str(), message.Value.length(),
+         mDocument.GetAllocator()),
+      mDocument.GetAllocator());
+
+   if (!message.Params.empty())
+   {
+      rapidjson::Value paramsArray(rapidjson::kArrayType);
+
+      for (const std::string& param : message.Params)
+      {
+         paramsArray.PushBack(
+            rapidjson::Value(
+               param.c_str(), param.length(), mDocument.GetAllocator()),
+            mDocument.GetAllocator());
+      }
+
+      messageObject.AddMember(
+         "params", std::move(paramsArray), mDocument.GetAllocator());
+   }
+
+   mDocument.AddMember(
+      "message", std::move(messageObject), mDocument.GetAllocator());
+}
+
+void Report::ReportImpl::AddUserComment(const std::string& message)
+{
+   // We only allow adding comment to exceptions now
+   if (!mDocument.HasMember("exception") || message.empty())
+      return;
+
+   rapidjson::Value& topException = mDocument["exception"]["values"][0];
+
+   if (!topException.IsObject())
+      return;
+
+   rapidjson::Value& mechanism = topException["mechanism"];
+
+   // Create a data object if it still does not exist
+   if (!mechanism.HasMember("data"))
+   {
+      mechanism.AddMember(
+         "data", rapidjson::Value(rapidjson::kObjectType),
+         mDocument.GetAllocator());
+   }
+
+   // Add a comment itself
+   mechanism["data"].AddMember(
+      "user_comment",
+      rapidjson::Value(
+         message.data(), message.length(), mDocument.GetAllocator()),
+      mDocument.GetAllocator());
+}
+
+
+void Report::ReportImpl::Send(CompletionHandler completionHandler) const
+{
+   const std::string serializedDocument = ToString(false);
+
+   network_manager::Request request =
+      SentryRequestBuilder::Get().CreateRequest();
+
+   auto response = network_manager::NetworkManager::GetInstance().doPost(
+      request, serializedDocument.data(), serializedDocument.size());
+
+   response->setRequestFinishedCallback(
+      [response, handler = std::move(completionHandler)](network_manager::IResponse*) {
+         const std::string responseData = response->readAll<std::string>();
+
+         wxLogDebug(responseData.c_str());
+
+         if (handler)
+            handler(response->getHTTPCode(), responseData);
+      });
+}
+
+std::string Report::ReportImpl::ToString(bool pretty) const
+{
+   rapidjson::StringBuffer buffer;
+
+   if (pretty)
+   {
+      rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(buffer);
+      mDocument.Accept(writer);
+   }
+   else
+   {
+      rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+      mDocument.Accept(writer);
+   }
+
+   return std::string(buffer.GetString());
+}
+
+// Some versions of GCC won't handle defaulted destructor correctly, 
+// generating error about std::unqiue_ptr destructor for a forward decalraed class
```

Comment:
spelling

Reply:
For Clang and CL it is sufficient to write `Report::~Report() = default;` However, I can't find which behavior is correct in C++ standard. I will remove the comment